lib/THC/THCGeneral.c

```diff
@@ -42,21 +101,153 @@ void THCudaInit(THCState* state)
+      newStreams[dev][stream] = state->streamsPerDevice[dev][stream];
+    }
+
+    /* Allocate new streams */
+    for (int stream = state->numUserStreams + 1; stream <= numStreams; ++stream) {
+      newStreams[dev][stream] = NULL;
+      THCudaCheck(cudaStreamCreate(&newStreams[dev][stream]));
+    }
+  }
+
+  cudaStream_t** oldStreams = state->streamsPerDevice;
+  state->streamsPerDevice = newStreams;
+  state->numUserStreams = numStreams;
+
+  for (int dev = 0; dev < state->numDevices; ++dev) {
+    free(oldStreams[dev]);
+  }
+  free(oldStreams);
+
+  THCudaCheck(cudaSetDevice(prevDev));
+}
+
+int THCState_getNumStreams(THCState* state)
+{
+  return state->numUserStreams;
+}
+
+void THCState_resetStreams(THCState* state, int device)
+{
+  if (state->currentStream !=
+      state->streamsPerDevice[device][state->currentPerDeviceStream]) {
+    THError("Unexpected stream state");
+  }
+
+  /* Reallocate all streams for the current device; the 0 stream
+     doesn't need updating */
+  for (int dev = 0; dev < state->numDevices; ++dev) {
+    for (int stream = 1; stream <= state->numUserStreams; ++stream) {
+      THCudaCheck(cudaStreamCreate(&state->streamsPerDevice[dev][stream]));
+    }
+  }
+
+  state->currentStream =
+    state->streamsPerDevice[device][state->currentPerDeviceStream];
+}
+
+cudaStream_t THCState_getDeviceStream(THCState *state, int device, int stream)
+{
+  /* `device` is a CUDA index */
+  if (device >= state->numDevices || device < 0)
+  {
+    THError("%d is not a device", device + 1 /* back to Torch index */);
+  }
+
+  /* Stream 0 is the default stream, 1 ... `numUserStreams` are Torch streams */
+  if (stream > state->numUserStreams || stream < 0)
+  {
+    THError("%d is not a stream", stream);
+  }
+
+  return state->streamsPerDevice[device][stream];
+}
+
+cudaStream_t THCState_getCurrentStream(THCState *state)
+{
+  /* This is called at the point of kernel execution.
+     For some debugging code or improperly instrumented kernels,
+     `state` is null */
+  if (state) {
+    return state->currentStream;
+  } else {
+    /* assume default stream */
+    return NULL;
+  }
+}
+
+int THCState_getCurrentStreamIndex(THCState *state)
+{
+  return state->currentPerDeviceStream;
+}
+
+void THCState_setStream(THCState *state, int device, int stream)
+{
+  /* `device` is a CUDA index */
+  if (device >= state->numDevices || device < 0)
+  {
+    THError("%d is not a device", device + 1 /* back to Torch index */);
+  }
+
+  /* Stream 0 is the default stream, 1 ... `numUserStreams` are Torch streams */
+  if (stream > state->numUserStreams || stream < 0)
+  {
+    THError("%d is not a stream", stream);
+  }
+
+  state->currentStream = state->streamsPerDevice[device][stream];
+  state->currentPerDeviceStream = stream;
+  THCudaBlas_setStream(state, device, state->currentStream);
+}
+
+void THCState_setStreamForCurrentDevice(THCState *state, int stream)
+{
+  if (state->currentPerDeviceStream != stream)
```

Comment:
cublasSetStream is relative to the cublasHandle_t, not the CPU thread. cudaSetDevice is thread-local state, so the other thread must call to set that when it comes into scope.

Crossing CPU threads with the same Lua state seems dangerous to me anyways, since there's no explicit memory barrier anywhere in Torch/cutorch; only whatever mutual exclusion device you are using must provide the memory barrier, but in some cases the compiler must see the memory barrier too in order to generate proper code, especially if you are within a function (e.g., write something out, expect some other thread to pick it up without a memory barrier in the way; there's no explicit guarantee in the C11/C++11 memory model that writes must be made visible to other threads without such a barrier.)